observability/logging.go

```diff
@@ -0,0 +1,232 @@
+		}
+		grpcLogRecord.PayloadTruncated = binlogEntry.GetPayloadTruncated()
+		setPeerIfPresent(binlogEntry, grpcLogRecord)
+	case binlogpb.GrpcLogEntry_EVENT_TYPE_SERVER_HEADER:
+		grpcLogRecord.EventType = grpclogrecordpb.GrpcLogRecord_GRPC_CALL_RESPONSE_HEADER
+		grpcLogRecord.Metadata = translateMetadata(binlogEntry.GetServerHeader().Metadata)
+		grpcLogRecord.PayloadTruncated = binlogEntry.GetPayloadTruncated()
+		setPeerIfPresent(binlogEntry, grpcLogRecord)
+	case binlogpb.GrpcLogEntry_EVENT_TYPE_CLIENT_MESSAGE:
+		grpcLogRecord.EventType = grpclogrecordpb.GrpcLogRecord_GRPC_CALL_REQUEST_MESSAGE
+		grpcLogRecord.Message = binlogEntry.GetMessage().GetData()
+		grpcLogRecord.PayloadSize = binlogEntry.GetMessage().GetLength()
+		grpcLogRecord.PayloadTruncated = binlogEntry.GetPayloadTruncated()
+	case binlogpb.GrpcLogEntry_EVENT_TYPE_SERVER_MESSAGE:
+		grpcLogRecord.EventType = grpclogrecordpb.GrpcLogRecord_GRPC_CALL_RESPONSE_MESSAGE
+		grpcLogRecord.Message = binlogEntry.GetMessage().GetData()
+		grpcLogRecord.PayloadSize = binlogEntry.GetMessage().GetLength()
+		grpcLogRecord.PayloadTruncated = binlogEntry.GetPayloadTruncated()
+	case binlogpb.GrpcLogEntry_EVENT_TYPE_CLIENT_HALF_CLOSE:
+		grpcLogRecord.EventType = grpclogrecordpb.GrpcLogRecord_GRPC_CALL_HALF_CLOSE
+	case binlogpb.GrpcLogEntry_EVENT_TYPE_SERVER_TRAILER:
+		grpcLogRecord.EventType = grpclogrecordpb.GrpcLogRecord_GRPC_CALL_TRAILER
+		grpcLogRecord.Metadata = translateMetadata(binlogEntry.GetTrailer().Metadata)
+		grpcLogRecord.StatusCode = binlogEntry.GetTrailer().GetStatusCode()
+		grpcLogRecord.StatusMessage = binlogEntry.GetTrailer().GetStatusMessage()
+		grpcLogRecord.StatusDetails = binlogEntry.GetTrailer().GetStatusDetails()
+		grpcLogRecord.PayloadTruncated = binlogEntry.GetPayloadTruncated()
+		setPeerIfPresent(binlogEntry, grpcLogRecord)
+	case binlogpb.GrpcLogEntry_EVENT_TYPE_CANCEL:
+		grpcLogRecord.EventType = grpclogrecordpb.GrpcLogRecord_GRPC_CALL_CANCEL
+	default:
+		logger.Warningf("unknown event type: %v", binlogEntry.Type)
+		return
+	}
+	grpcLogRecord.ServiceName = ml.serviceName
+	grpcLogRecord.MethodName = ml.methodName
+	// CloudLogging client doesn't return error on entry write. Entry writes
+	// don't mean the data will be uploaded immediately.
+	globalLoggingExporter.EmitGrpcLogRecord(grpcLogRecord)
+}
+
+type observabilityBinaryLogger struct {
+	// originalLogger is needed to ensure binary logging users won't be impacted
+	// by this plugin. Users are allowed to subscribe to a completely different
+	// set of methods.
+	originalLogger iblog.Logger
+	// wrappedLogger is needed to reuse the control string parsing, logger
+	// config managing logic
+	wrappedLogger iblog.Logger
+}
+
+func (l *observabilityBinaryLogger) GetMethodLogger(methodName string) iblog.MethodLogger {
+	var ol, ml iblog.MethodLogger
+	if l.originalLogger != nil {
+		ol = l.originalLogger.GetMethodLogger(methodName)
+	}
+	if l.wrappedLogger == nil {
```

Comment:
If a gRPC client _is_ started in an init, then this will be a data race.  Best to read this field atomically.